lib/chainweb/cut.ex

```diff
@@ -0,0 +1,71 @@
+defmodule Kadena.Chainweb.Cut do
+  @moduledoc """
+  `Cut` struct definition.
+  """
+  @behaviour Kadena.Chainweb.Type
+
+  @type hashes :: map()
+  @type height :: non_neg_integer()
+  @type id :: String.t()
+  @type instance :: String.t()
+  @type origin :: map() | nil
+  @type weight :: String.t()
+  @type chain_id :: non_neg_integer()
+  @type error :: {:error, Keyword.t()}
+  @type result :: t() | error()
+
+  @type t :: %__MODULE__{
+          hashes: hashes(),
+          height: height(),
+          weight: weight(),
+          origin: origin(),
+          id: id(),
+          instance: instance()
+        }
+
+  defstruct [:height, :weight, :id, :instance, :origin, hashes: %{}]
+
+  @impl true
+  def new(args \\ [])
+  def new(attrs) when is_map(attrs), do: struct(%__MODULE__{}, attrs)
+  def new(attrs) when is_list(attrs), do: struct(%__MODULE__{}, attrs)
+  def new(_attrs), do: {:error, [args: :invalid_format]}
+
+  @spec set_hashes(cut :: t(), hashes :: hashes()) :: result()
+  def set_hashes(%__MODULE__{} = cut, %{} = hashes), do: %{cut | hashes: hashes}
+  def set_hashes(%__MODULE__{}, _hashes), do: {:error, [hashes: :not_a_map]}
```

Comment:
Here we can call the `add_hash` function iterating the hashes map to perform the validations per key-value!